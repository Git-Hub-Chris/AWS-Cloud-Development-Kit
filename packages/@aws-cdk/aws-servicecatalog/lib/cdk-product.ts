import * as core from '@aws-cdk/core';
import { Product } from './product';

/**
 * Properties for CDKProduct.
 */
export interface CDKProductProps {
  readonly productName: string;
  readonly stackName: string;
  readonly owner: string;
  readonly description?: string;
  readonly artifaceName?: string;
  readonly app: core.App;
}

export class CDKProduct extends Product {
  public readonly productName: string;
  constructor(scope: core.Construct, id: string, props: CDKProductProps) {
    const stackArtifact = props.app.synth().getStackByName(props.stackName);
    if (stackArtifact.assets.length !== 0) {
      throw new Error('CDKProduct - Stack cannot have assets');
    }

    super(scope, id, {
      productName: props.productName,
      templatePath: stackArtifact.templateFullPath,
      owner: props.owner,
      description: props.description,
      artifactName: props.artifaceName,
    });
    this.productName = props.productName;
  }
}