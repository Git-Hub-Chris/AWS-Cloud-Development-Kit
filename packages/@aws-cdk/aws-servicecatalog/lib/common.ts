/**
 * Constant for the context directory to store retained ProductStack templates.
 */
export const PRODUCT_STACK_CONTEXT_DIRECTORY = 'product-stack-context';

/**
 * The language code.
 * Used for error and logging messages for end users.
 * The default behavior if not specified is English.
 */
export enum MessageLanguage {
  /**
   * English
   */
  EN = 'en',

  /**
   * Japanese
   */
  JP = 'jp',

  /**
   * Chinese
   */
  ZH = 'zh'
}

/**
 * The types of cloudFormationTemplates for product versions.
 * Used to determine the source of a cloudFormationTemplate and apply logic accordingly
 */
export enum TemplateType {
  /**
   * AssetTemplate
   */
  ASSET = 'AssetTemplate',

  /**
   * UrlTemplate
   */
  URL = 'UrlTemplate',

  /**
   * ProductStackTemplate
   */
  PRODUCT_STACK = 'ProductStackTemplate',

  /**
   * ProductStackContextTemplate
   */
  PRODUCT_STACK_CONTEXT = 'ProductStackContextTemplate'
}

/**
   * The strategy to use for a ProductStack deployment.
   * Determines how a productVersion is saved and deployed.
   */
export enum VersioningStrategy {
  /**
   * Default Strategy for ProductStack deployment.
   * This strategy will overwrite existing versions when deployed.
   */
  DEFAULT = 'Default',

  /**
   * Retain previously deployed ProductStacks in a local context directory.
   * This strategy will not overwrite existing versions when deployed.
   */
  RETAIN_PREVIOUS_VERSIONS = 'RetainPreviousVersions',
}

/**
 * A wrapper class containing useful metadata about the product version
 */
export class ProductVersionDetails {
  /**
   * Unique id denoting the stack path of a product.
   */
  public productPathUniqueId?: string;

  /**
   * Id of a ProductStack
   */
  public productStackId?: string;

  /**
   * Name of a productVersion
   */
  public productVersionName?: string;

  /**
   * Versioning strategy to use for a ProductStack deployment.
   */
  public versioningStrategy?: VersioningStrategy
}
