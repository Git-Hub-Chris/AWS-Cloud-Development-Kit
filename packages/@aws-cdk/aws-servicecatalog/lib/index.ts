// AWS::ServiceCatalog CloudFormation Resources:
export * from './servicecatalog.generated';
export * from './portfolio';
export * from './product';
export * from './cdk-product';
export * from './language';
export * from './constraints';