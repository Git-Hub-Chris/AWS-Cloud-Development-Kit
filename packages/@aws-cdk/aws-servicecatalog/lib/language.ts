/**
 * AWS Service Catalog Languages
 */
export enum Language {
  /**
   * English
   */
  ENGLISH = 'en',
  /**
   * Japanese
   */
  JAPANESE = 'jp',
  /**
   * Chinese
   */
  CHINESE = 'zh',
}