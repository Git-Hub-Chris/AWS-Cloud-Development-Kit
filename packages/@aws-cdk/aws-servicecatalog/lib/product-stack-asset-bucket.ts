import { Bucket, BucketProps } from '@aws-cdk/aws-s3';
import { BucketDeployment, ISource, Source } from '@aws-cdk/aws-s3-deployment';
import * as cdk from '@aws-cdk/core';
import { Construct, IConstruct } from 'constructs';

/**
 * Product stack asset bucket props.
 */
export interface ProductStackAssetBucketProps extends BucketProps {
}

/**
 * A Service Catalog product stack asset bucket, which is similar in form to an Amazon S3 bucket.
 * You can store multiple product stack assets and collectively deploy them to S3.
 */
export class ProductStackAssetBucket extends Bucket {
  private readonly assets: ISource[];

  constructor(scope: Construct, id: string, props: ProductStackAssetBucketProps = {}) {
    super(scope, id, props);

    if (props.bucketName == undefined) {
      throw new Error('BucketName must be defined for assetBucket');
    }

    this.assets = [];

    cdk.Aspects.of(this).add({
      visit(c: IConstruct) {
        if (c instanceof ProductStackAssetBucket) {
          c.deployAssets();
        };
      },
    });
  }

  /**
   * Asset are prepared for bulk deployment to S3.
   * @internal
   */
  public _addAsset(asset: cdk.FileAssetSource): void {
    const assetPath = './cdk.out/' + asset.fileName;
    this.assets.push(Source.asset(assetPath));
  }

  /**
   * Deploy all assets to S3.
   */
  private deployAssets() {
    if (this.assets.length > 0) {
      new BucketDeployment(this, 'AssetsBucketDeployment', {
        sources: this.assets,
        destinationBucket: this,
        extract: false,
        prune: false,
      });
    }
  }
}