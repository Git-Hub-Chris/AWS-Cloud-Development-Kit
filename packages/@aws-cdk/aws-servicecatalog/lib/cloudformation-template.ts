import * as s3_assets from '@aws-cdk/aws-s3-assets';
import { ProductVersionDetails, TemplateType } from './common';
import { hashValues } from './private/util';
import { ProductStack } from './product-stack';

// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
// eslint-disable-next-line no-duplicate-imports, import/order
import { Construct } from '@aws-cdk/core';

/**
 * Represents the Product Provisioning Artifact Template.
 */
export abstract class CloudFormationTemplate {
  /**
   * Template from URL
   * @param url The url that points to the provisioning artifacts template
   */
  public static fromUrl(url: string): CloudFormationTemplate {
    return new CloudFormationUrlTemplate(url);
  }

  /**
   * Loads the provisioning artifacts template from a local disk path.
   *
   * @param path A file containing the provisioning artifacts
   */
  public static fromAsset(path: string, options?: s3_assets.AssetOptions): CloudFormationTemplate {
    return new CloudFormationAssetTemplate(path, options);
  }

  /**
   * Creates a product with the resources defined in the given product stack.
   */
  public static fromProductStack(productStack: ProductStack): CloudFormationTemplate {
    return new CloudFormationProductStackTemplate(productStack);
  }

  /**
   * Creates a product with the resources defined in the given product stack and retains all previously deployed product stack versions.
   */
  public static fromProductStackHistory(productStack: ProductStack, locked: boolean, directory?: string ): CloudFormationTemplate {
    return new CloudFormationProductStackTemplate(productStack, locked, directory);
  }

  /**
   * Creates a product from a previously deployed product stack snapshot.
   */
  public static fromProductStackSnapshot(productStack: ProductStack, directory?: string ): CloudFormationTemplate {
    return new CloudFormationProductStackSnapshotTemplate(productStack, directory);
  }


  /**
   * Called when the product is initialized to allow this object to bind
   * to the stack, add resources and have fun.
   *
   * @param scope The binding scope. Don't be smart about trying to down-cast or
   * assume it's initialized. You may just use it as a construct scope.
   */
  public abstract bind(scope: Construct): CloudFormationTemplateConfig;
}

/**
 * Result of binding `Template` into a `Product`.
 */
export interface CloudFormationTemplateConfig {
  /**
    * The http url of the template in S3.
    */
  readonly httpUrl: string;
  /**
   * Additional metadata about the product version.
   * @default - No additional details provided
   */
  readonly productVersionDetails?: ProductVersionDetails;
  /**
   * The type of the template source.
   */
  readonly templateType: TemplateType;
}

/**
 * Template code from a Url.
 */
class CloudFormationUrlTemplate extends CloudFormationTemplate {
  constructor(private readonly url: string) {
    super();
  }

  public bind(_scope: Construct): CloudFormationTemplateConfig {
    return {
      httpUrl: this.url,
      templateType: TemplateType.URL,
    };
  }
}

/**
 * Template from a local file.
 */
class CloudFormationAssetTemplate extends CloudFormationTemplate {
  private asset?: s3_assets.Asset;

  /**
   * @param path The path to the asset file.
   */
  constructor(public readonly path: string, private readonly options: s3_assets.AssetOptions = { }) {
    super();
  }

  public bind(scope: Construct): CloudFormationTemplateConfig {
    // If the same AssetCode is used multiple times, retain only the first instantiation.
    if (!this.asset) {
      this.asset = new s3_assets.Asset(scope, `Template${hashValues(this.path)}`, {
        path: this.path,
        ...this.options,
      });
    }

    return {
      httpUrl: this.asset.httpUrl,
      templateType: TemplateType.ASSET,
    };
  }
}

/**
 * Template from a CDK defined product stack.
 */
class CloudFormationProductStackTemplate extends CloudFormationTemplate {
  /**
   * @param stack A service catalog product stack.
   */
  constructor(public readonly productStack: ProductStack,
    public readonly locked?: boolean, public readonly directory?: string) {
    super();
    const productVersionDetails = productStack._getProductVersionDetails();
    productVersionDetails.locked = this.locked;
    productVersionDetails.directory = this.directory;
  }

  public bind(_scope: Construct): CloudFormationTemplateConfig {
    return {
      httpUrl: this.productStack._getTemplateUrl(),
      productVersionDetails: this.productStack._getProductVersionDetails(),
      templateType: TemplateType.PRODUCT_STACK,
    };
  }
}

/**
 * Template from a previously deployed product stack.
 */
class CloudFormationProductStackSnapshotTemplate extends CloudFormationTemplate {
  /**
   * @param stack A service catalog product stack.
   */
  constructor(public readonly productStack: ProductStack, public readonly directory?: string) {
    super();
    const productVersionDetails = productStack._getProductVersionDetails();
    productVersionDetails.directory = this.directory;
  }

  public bind(_scope: Construct): CloudFormationTemplateConfig {
    return {
      httpUrl: this.productStack._getTemplateUrl(),
      productVersionDetails: this.productStack._getProductVersionDetails(),
      templateType: TemplateType.PRODUCT_STACK_SNAPSHOT,
    };
  }
}
