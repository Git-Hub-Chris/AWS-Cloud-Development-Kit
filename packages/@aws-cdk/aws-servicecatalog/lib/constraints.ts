import * as iam from '@aws-cdk/aws-iam';
import * as core from '@aws-cdk/core';
import * as constructs from 'constructs';
import { Language } from './language';
import { IPortfolio } from './portfolio';
import { IProduct } from './product';
import * as servicecatalog from './servicecatalog.generated';

export interface IConstraint extends core.IResource {
}

interface BaseConstraintProps {
  readonly acceptLanguage?: Language;
  readonly description?: string;
  readonly portfolio?: IPortfolio;
  readonly product?: IProduct;
}

export interface LaunchNotificationConstraintProps extends BaseConstraintProps {
  readonly notificationArns: string[];
}
export interface LaunchRoleConstraintProps extends BaseConstraintProps {
  readonly localRoleName?: string;
  readonly role?: iam.IRole;
}
export interface LaunchTemplateConstraintProps extends BaseConstraintProps {
  // TODO: make this it's own type
  readonly rules: string;
}
export interface ResourceUpdateConstraintProps extends BaseConstraintProps {
  readonly allowTagUpdateOnProvisionedProduct: boolean;
}
export interface StackSetConstraintProps extends BaseConstraintProps {
  readonly accounts: string[];
  readonly adminRole: iam.IRole;
  readonly executionRole: iam.IRole;
  readonly regions: string[];
  readonly allowStackInstanceControl: boolean;
}

abstract class ConstraintBase extends core.Resource implements IConstraint {
  readonly acceptLanguage?: Language;
  readonly description?: string;
  readonly portfolio?: IPortfolio;
  readonly product?: IProduct;

  constructor(scope: constructs.Construct, id: string, props: BaseConstraintProps) {
    super(scope, id);
    this.acceptLanguage = props.acceptLanguage;
    this.description = props.description;
    this.portfolio = props.portfolio;
    this.product = props.product;
  }
  // This function could potentially made to be more generic so that callers only need to pass the diff between base props and their specific props.
  protected absApply<R extends core.CfnResource, RP>(R: new (scope: core.Construct, id: string, props: RP) => R, baseProps: RP) {
    new R(this, 'productContraint', baseProps);
  }
}

export class LaunchNotificationConstraint extends ConstraintBase implements IConstraint {
  readonly notificationArns: string[];
  constructor(scope: constructs.Construct, id: string, props: LaunchNotificationConstraintProps) {
    super(scope, id, props);
    this.notificationArns = props.notificationArns;

    if (props.product && props.portfolio) {
      this.apply(props.product, props.portfolio);
    }
  }
  public apply(product: IProduct, portfolio: IPortfolio) {
    new servicecatalog.CfnLaunchNotificationConstraint(this, 'launchNotificationConstraint', {
      acceptLanguage: this.acceptLanguage,
      description: this.description,
      portfolioId: portfolio.portfolioId,
      productId: product.productId,
      notificationArns: this.notificationArns,
    });
  }
}
export class LaunchRoleConstraint extends ConstraintBase implements IConstraint {
  readonly localRoleName?: string;
  readonly role?: iam.IRole;
  constructor(scope: constructs.Construct, id: string, props: LaunchRoleConstraintProps) {
    super(scope, id, props);
    this.localRoleName = props.localRoleName;
    this.role = props.role;

    if (props.product && props.portfolio) {
      this.apply(props.product, props.portfolio);
    }
  }
  public apply(product: IProduct, portfolio: IPortfolio) {
    new servicecatalog.CfnLaunchRoleConstraint(this, 'launchRoleConstraint', {
      acceptLanguage: this.acceptLanguage,
      description: this.description,
      portfolioId: portfolio.portfolioId,
      productId: product.productId,
      localRoleName: this.localRoleName,
      roleArn: this.role?.roleArn,
    });
  }
}
export class LaunchTemplateConstraint extends ConstraintBase implements IConstraint {
  readonly rules: string;
  constructor(scope: constructs.Construct, id: string, props: LaunchTemplateConstraintProps) {
    super(scope, id, props);
    this.rules = props.rules;

    if (props.product && props.portfolio) {
      this.apply(props.product, props.portfolio);
    }
  }
  public apply(product: IProduct, portfolio: IPortfolio) {
    new servicecatalog.CfnLaunchTemplateConstraint(this, 'launchTemplateConstraint', {
      acceptLanguage: this.acceptLanguage,
      description: this.description,
      portfolioId: portfolio.portfolioId,
      productId: product.productId,
      rules: this.rules,
    });
  }
}
export class ResourceUpdateConstraint extends ConstraintBase implements IConstraint {
  readonly allowTagUpdateOnProvisionedProduct: boolean;
  private tagUpdateOnProvisionedProduct: 'ALLOWED' | 'NOT_ALLOWED'

  constructor(scope: constructs.Construct, id: string, props: ResourceUpdateConstraintProps) {
    super(scope, id, props);
    this.allowTagUpdateOnProvisionedProduct = props.allowTagUpdateOnProvisionedProduct;
    if (this.allowTagUpdateOnProvisionedProduct) {
      this.tagUpdateOnProvisionedProduct = 'ALLOWED';
    } else {
      this.tagUpdateOnProvisionedProduct = 'NOT_ALLOWED';
    }

    if (props.product && props.portfolio) {
      this.apply(props.product, props.portfolio);
    }
  }
  public apply(product: IProduct, portfolio: IPortfolio) {
    new servicecatalog.CfnResourceUpdateConstraint(this, 'resourceUpdateConstraint', {
      acceptLanguage: this.acceptLanguage,
      description: this.description,
      portfolioId: portfolio.portfolioId,
      productId: product.productId,
      tagUpdateOnProvisionedProduct: this.tagUpdateOnProvisionedProduct,
    });
  }
}

export class StackSetConstraint extends ConstraintBase implements IConstraint {
  readonly accounts: string[];
  readonly adminRole: iam.IRole;
  readonly executionRole: iam.IRole;
  readonly regions: string[];
  readonly allowStackInstanceControl: boolean;
  private stackInstanceControl: 'ALLOWED' | 'NOT_ALLOWED';

  constructor(scope: constructs.Construct, id: string, props: StackSetConstraintProps) {
    super(scope, id, props);
    this.accounts = props.accounts;
    this.adminRole = props.adminRole;
    this.executionRole = props.executionRole;
    this.regions = props.regions;
    this.allowStackInstanceControl = props.allowStackInstanceControl;

    // As of 2020-10-23 the CloudFormation Resource Specification says the "Description" parameter of this resource is required.
    // This is likely a bug and this field can be removed once this issue is resolved.
    this.description === undefined ? props.description : 'No description';

    if (this.allowStackInstanceControl) {
      this.stackInstanceControl = 'ALLOWED';
    } else {
      this.stackInstanceControl = 'NOT_ALLOWED';
    }
    if (props.product && props.portfolio) {
      this.apply(props.product, props.portfolio);
    }
  }
  public apply(product: IProduct, portfolio: IPortfolio) {
    new servicecatalog.CfnStackSetConstraint(this, 'stackSetConstraint', {
      acceptLanguage: this.acceptLanguage,
      description: this.description!,
      portfolioId: portfolio.portfolioId,
      productId: product.productId,
      stackInstanceControl: this.stackInstanceControl,
      accountList: this.accounts,
      adminRole: this.adminRole.roleArn,
      regionList: this.regions,
      executionRole: this.executionRole.roleArn,
    });
  }
}
