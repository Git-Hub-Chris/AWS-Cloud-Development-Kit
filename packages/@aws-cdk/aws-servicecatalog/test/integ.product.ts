import * as path from 'path';
import * as s3 from '@aws-cdk/aws-s3';
import * as s3_assets from '@aws-cdk/aws-s3-assets';
import * as sns from '@aws-cdk/aws-sns';
import * as cdk from '@aws-cdk/core';
import * as servicecatalog from '../lib';
import { ProductStackHistory, ProductStackProps } from '../lib';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'integ-servicecatalog-product');

class TestProductStack extends servicecatalog.ProductStack {
  constructor(scope: any, id: string) {
    super(scope, id);

    new sns.Topic(this, 'TopicProduct');
  }
}

class TestAssetProductStack extends servicecatalog.ProductStack {
  constructor(scope: any, id: string, props?: ProductStackProps) {
    super(scope, id, props);

    new s3_assets.Asset(this, 'testAsset', {
      path: path.join(__dirname, 'products.template.zip'),
    });
  }
}

const productStackHistory = new ProductStackHistory(stack, 'ProductStackHistory', {
  productStack: new TestProductStack(stack, 'SNSTopicProduct3'),
  currentVersionName: 'v1',
  currentVersionLocked: true,
});

const testAssetBucket = new s3.Bucket(stack, 'TestAssetBucket', {
  bucketName: 'product-stack-asset-bucket-12345678-test-region',
  removalPolicy: cdk.RemovalPolicy.DESTROY,
  autoDeleteObjects: true,
});

const product = new servicecatalog.CloudFormationProduct(stack, 'TestProduct', {
  productName: 'testProduct',
  owner: 'testOwner',
  productVersions: [
    {
      validateTemplate: false,
      cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromUrl(
        'https://awsdocs.s3.amazonaws.com/servicecatalog/development-environment.template'),
    },
    {
      cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromAsset(path.join(__dirname, 'product1.template.json')),
    },
    {
      cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromAsset(path.join(__dirname, 'product2.template.json')),
    },
    {
      cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromProductStack(new TestProductStack(stack, 'SNSTopicProduct1')),
    },
    {
      cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromProductStack(new TestProductStack(stack, 'SNSTopicProduct2')),
    },
    {
      validateTemplate: false,
      cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromProductStack(new TestAssetProductStack(stack, 'S3AssetProduct', {
        assetBucket: testAssetBucket,
      })),
    },
    productStackHistory.currentVersion(),
  ],
});

const tagOptions = new servicecatalog.TagOptions(stack, 'TagOptions', {
  allowedValuesForTags: {
    key1: ['value1', 'value2'],
    key2: ['value1'],
  },
});

product.associateTagOptions(tagOptions);

app.synth();
