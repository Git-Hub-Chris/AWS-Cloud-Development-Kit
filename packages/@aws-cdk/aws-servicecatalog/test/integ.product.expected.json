{
  "Resources": {
    "TestProduct7606930B": {
      "Type": "AWS::ServiceCatalog::CloudFormationProduct",
      "Properties": {
        "Name": "testProduct",
        "Owner": "testOwner",
        "ProvisioningArtifactParameters": [
          {
            "DisableTemplateValidation": true,
            "Info": {
              "LoadTemplateFromURL": "https://awsdocs.s3.amazonaws.com/servicecatalog/development-environment.template"
            }
          },
          {
            "DisableTemplateValidation": false,
            "Info": {
              "LoadTemplateFromURL": {
                "Fn::Join": [
                  "",
                  [
                    "https://s3.",
                    {
                      "Ref": "AWS::Region"
                    },
                    ".",
                    {
                      "Ref": "AWS::URLSuffix"
                    },
                    "/",
                    {
                      "Ref": "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eS3BucketCF629374"
                    },
                    "/",
                    {
                      "Fn::Select": [
                        0,
                        {
                          "Fn::Split": [
                            "||",
                            {
                              "Ref": "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eS3VersionKeyC94BC7EE"
                            }
                          ]
                        }
                      ]
                    },
                    {
                      "Fn::Select": [
                        1,
                        {
                          "Fn::Split": [
                            "||",
                            {
                              "Ref": "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eS3VersionKeyC94BC7EE"
                            }
                          ]
                        }
                      ]
                    }
                  ]
                ]
              }
            }
          },
          {
            "DisableTemplateValidation": false,
            "Info": {
              "LoadTemplateFromURL": {
                "Fn::Join": [
                  "",
                  [
                    "https://s3.",
                    {
                      "Ref": "AWS::Region"
                    },
                    ".",
                    {
                      "Ref": "AWS::URLSuffix"
                    },
                    "/",
                    {
                      "Ref": "AssetParameters6412a5f4524c6b41d26fbeee226c68c2dad735393940a51008d77e6f8b1038f5S3Bucket85C3FF42"
                    },
                    "/",
                    {
                      "Fn::Select": [
                        0,
                        {
                          "Fn::Split": [
                            "||",
                            {
                              "Ref": "AssetParameters6412a5f4524c6b41d26fbeee226c68c2dad735393940a51008d77e6f8b1038f5S3VersionKey34A02667"
                            }
                          ]
                        }
                      ]
                    },
                    {
                      "Fn::Select": [
                        1,
                        {
                          "Fn::Split": [
                            "||",
                            {
                              "Ref": "AssetParameters6412a5f4524c6b41d26fbeee226c68c2dad735393940a51008d77e6f8b1038f5S3VersionKey34A02667"
                            }
                          ]
                        }
                      ]
                    }
                  ]
                ]
              }
            }
          },
          {
            "DisableTemplateValidation": false,
            "Info": {
              "LoadTemplateFromURL": {
                "Fn::Join": [
                  "",
                  [
                    "https://s3.",
                    {
                      "Ref": "AWS::Region"
                    },
                    ".",
                    {
                      "Ref": "AWS::URLSuffix"
                    },
                    "/",
                    {
                      "Ref": "AssetParameters08d08044426323f29a1dec4ad6ba40a16a9b2c8695f4408c833c01a7e536b8f2S3Bucket971225BF"
                    },
                    "/",
                    {
                      "Fn::Select": [
                        0,
                        {
                          "Fn::Split": [
                            "||",
                            {
                              "Ref": "AssetParameters08d08044426323f29a1dec4ad6ba40a16a9b2c8695f4408c833c01a7e536b8f2S3VersionKeyF3C21C22"
                            }
                          ]
                        }
                      ]
                    },
                    {
                      "Fn::Select": [
                        1,
                        {
                          "Fn::Split": [
                            "||",
                            {
                              "Ref": "AssetParameters08d08044426323f29a1dec4ad6ba40a16a9b2c8695f4408c833c01a7e536b8f2S3VersionKeyF3C21C22"
                            }
                          ]
                        }
                      ]
                    }
                  ]
                ]
              }
            }
          }
        ]
      }
    }
  },
  "Parameters": {
    "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eS3BucketCF629374": {
      "Type": "String",
      "Description": "S3 bucket for asset \"b59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09e\""
    },
    "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eS3VersionKeyC94BC7EE": {
      "Type": "String",
      "Description": "S3 key for asset version \"b59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09e\""
    },
    "AssetParametersb59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09eArtifactHashB9EF04B2": {
      "Type": "String",
      "Description": "Artifact hash for asset \"b59f768286e16b69628bb23b9c1a1f07300a24101b8979d8e2a94ff1ab03d09e\""
    },
    "AssetParameters6412a5f4524c6b41d26fbeee226c68c2dad735393940a51008d77e6f8b1038f5S3Bucket85C3FF42": {
      "Type": "String",
      "Description": "S3 bucket for asset \"6412a5f4524c6b41d26fbeee226c68c2dad735393940a51008d77e6f8b1038f5\""
    },
    "AssetParameters6412a5f4524c6b41d26fbeee226c68c2dad735393940a51008d77e6f8b1038f5S3VersionKey34A02667": {
      "Type": "String",
      "Description": "S3 key for asset version \"6412a5f4524c6b41d26fbeee226c68c2dad735393940a51008d77e6f8b1038f5\""
    },
    "AssetParameters6412a5f4524c6b41d26fbeee226c68c2dad735393940a51008d77e6f8b1038f5ArtifactHashDC26AFAC": {
      "Type": "String",
      "Description": "Artifact hash for asset \"6412a5f4524c6b41d26fbeee226c68c2dad735393940a51008d77e6f8b1038f5\""
    },
    "AssetParameters08d08044426323f29a1dec4ad6ba40a16a9b2c8695f4408c833c01a7e536b8f2S3Bucket971225BF": {
      "Type": "String",
      "Description": "S3 bucket for asset \"08d08044426323f29a1dec4ad6ba40a16a9b2c8695f4408c833c01a7e536b8f2\""
    },
    "AssetParameters08d08044426323f29a1dec4ad6ba40a16a9b2c8695f4408c833c01a7e536b8f2S3VersionKeyF3C21C22": {
      "Type": "String",
      "Description": "S3 key for asset version \"08d08044426323f29a1dec4ad6ba40a16a9b2c8695f4408c833c01a7e536b8f2\""
    },
    "AssetParameters08d08044426323f29a1dec4ad6ba40a16a9b2c8695f4408c833c01a7e536b8f2ArtifactHash0F048917": {
      "Type": "String",
      "Description": "Artifact hash for asset \"08d08044426323f29a1dec4ad6ba40a16a9b2c8695f4408c833c01a7e536b8f2\""
    }
  }
}